akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -1138,6 +1175,13 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef, joinConfigCompatCh
         shutdownSelfWhenDown()
       }
 
+      if (latestGossip.members.exists(m => MembershipState.shutdownStates(m.status)) && (MembershipState
+            .allowedToPrepareToShutdown(selfStatus))) {
+        // TODO switch to debug
+        logInfo("Detected full cluster shutdown")
+        self ! ClusterUserAction.PrepareForShutdown
```

Comment:
Since we change all in `def startShutdown` it's less important to do this immediately when receiving gossip. We still need to check because there can be a transition on another node (Joining -> Up). It's not great for performance to add another full latestGossip.members iteration in each receiveGossip.

I suggest that we move this check to `def leaderActions`. We have some other background tasks there, such as `cleanupExitingConfirmed`.

Reply:
Yep good idea now we only rely on this for edge cases

Reply:
And will switch to check if the current node is in a state that can prepare to shut down to avoid the iteration most of the time